File: app/lib/shared/widgets/member_avatar.dart
import 'package:cached_network_image/cached_network_image.dart';
import 'package:flutter/material.dart';
import 'package:sbc_contacts/core/theme/sbc_colors.dart';

/// Avatar with network image + initials fallback (cached, fade-in).
///
/// The fallback is a saturated tile rather than a grey placeholder: most members
/// have no photo, so this is the common case, not the exception. The colour is
/// derived from the initials, which means the same member keeps the same colour
/// on every screen and a scanned list gains a second, non-textual cue.
class MemberAvatar extends StatelessWidget {
  const MemberAvatar({
    required this.initials,
    this.avatarUrl,
    this.radius = 24,
    this.rounded = false,
    super.key,
  });

  final String initials;
  final String? avatarUrl;
  final double radius;

  /// Draw as a rounded square instead of a circle. Used in the directory
  /// lists, where the squared tile lines up with the card's own corners.
  final bool rounded;

  /// Deliberately short and well separated in hue; all of them carry white
  /// text at the weight used below.
  static const _palette = <Color>[
    SbcColors.primary,
    SbcColors.secondaryDark,
    SbcColors.accentDark,
    Color(0xFF5B5BD6), // indigo
    Color(0xFF0E9384), // teal
    Color(0xFFD1495B), // coral
    Color(0xFF7A5AF8), // violet
    Color(0xFFEA7317), // amber
  ];

  static Color colorFor(String seed) {
    if (seed.isEmpty) return _palette.first;
    var hash = 0;
    for (final unit in seed.codeUnits) {
      hash = (hash * 31 + unit) & 0x7fffffff;
    }
    return _palette[hash % _palette.length];
  }

  double get _cornerRadius => rounded ? radius * 0.62 : radius;

  @override
  Widget build(BuildContext context) {
    final bg = colorFor(initials);
    final fallback = Container(
      width: radius * 2,
      height: radius * 2,
      alignment: Alignment.center,
      decoration: BoxDecoration(
        color: bg,
        borderRadius: BorderRadius.circular(_cornerRadius),
      ),
      child: Text(
        initials,
        style: TextStyle(
          color: Colors.white,
          fontWeight: FontWeight.w800,
          fontSize: radius * 0.72,
          letterSpacing: -0.5,
        ),
      ),
    );
    if (avatarUrl == null || avatarUrl!.isEmpty) return fallback;
    return ClipRRect(
      borderRadius: BorderRadius.circular(_cornerRadius),
      child: CachedNetworkImage(
        imageUrl: avatarUrl!,
        width: radius * 2,
        height: radius * 2,
        fit: BoxFit.cover,
        placeholder: (_, __) => fallback,
        errorWidget: (_, __, ___) => fallback,
      ),
    );
  }
}

File: ios/SBCContacts/Shared/Components/MemberAvatar.swift
import SwiftUI

/// Avatar with network image + initials fallback.
///
/// The fallback is a saturated tile, not a grey placeholder: most members have
/// no photo, so it is the common case. Its colour is derived from the initials,
/// so a member keeps the same colour on every screen and a scanned list gains a
/// second, non-textual cue.
struct MemberAvatar: View {
    let initials: String
    var avatarUrl: String?
    var radius: CGFloat = 24
    /// A rounded square instead of a circle (directory lists).
    var rounded = false

    private static let palette: [Color] = [
        SBCColors.primary,
        SBCColors.secondaryDark,
        SBCColors.accentDark,
        Color(hex: 0x5B5BD6), // indigo
        Color(hex: 0x0E9384), // teal
        Color(hex: 0xD1495B), // coral
        Color(hex: 0x7A5AF8), // violet
        Color(hex: 0xEA7317), // amber
    ]

    static func color(for seed: String) -> Color {
        guard !seed.isEmpty else { return palette[0] }
        var hash = 0
        for unit in seed.utf16 {
            hash = (hash &* 31 &+ Int(unit)) & 0x7FFF_FFFF
        }
        return palette[hash % palette.count]
    }

    private var shape: AnyShape {
        rounded
            ? AnyShape(RoundedRectangle(cornerRadius: radius * 0.62, style: .continuous))
            : AnyShape(Circle())
    }

    var body: some View {
        Group {
            if let avatarUrl, !avatarUrl.isEmpty, let url = URL(string: avatarUrl) {
                AsyncImage(url: url, transaction: Transaction(animation: .easeOut(duration: 0.2))) { phase in
                    if let image = phase.image {
                        image.resizable().scaledToFill()
                    } else {
                        fallback
                    }
                }
            } else {
                fallback
            }
        }
        .frame(width: radius * 2, height: radius * 2)
        .clipShape(shape)
        .accessibilityHidden(true)
    }

    private var fallback: some View {
        ZStack {
            Self.color(for: initials)
            Text(initials)
                .font(SBCFontFixed.font(size: radius * 0.72, weight: .heavy))
                .tracking(-0.5)
                .foregroundStyle(.white)
        }
    }
}

/// Fixed-size Montserrat, for glyphs sized from their container (not Dynamic Type).
enum SBCFontFixed {
    static func bold(size: CGFloat) -> Font {
        Font(SBCFont.uiFont(size: size, weight: .bold))
    }

    static func font(size: CGFloat, weight: Font.Weight) -> Font {
        Font(SBCFont.uiFont(size: size, weight: weight))
    }
}

/// Up to two initials from a display name ("Jean Paul Mbarga" → "JP").
func initials(of name: String) -> String {
    let parts = name.split(whereSeparator: \.isWhitespace)
    guard let first = parts.first?.first else { return "?" }
    guard parts.count > 1, let second = parts[1].first else { return String(first).uppercased() }
    return (String(first) + String(second)).uppercased()
}
